DOCS-5805-flatten-duke cookbook (#1202)
mulesoft/docs-mule-runtime-3: modules/ROOT/pages/dataweave-cookbook-flatten-arrays.adoc

= Flatten Elements of Arrays
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:keywords: studio, anypoint, transform, transformer, format, aggregate, rename, split, filter convert, xml, json, csv, pojo, java object, metadata, dataweave, data weave, datamapper, dwl, dfl, dw, output structure, input structure, map, mapping

> DataWeave can flatten subarrays of an array and collections of key-value pairs

Reviewer: I wonder if this should be worded to address this as a user goal or task, and why?
Something like, "You can do this using DataWeave, and here is why you might want or need to do it."

Reviewer: Also, do you think that this existing intro sentence prepares your audience for the three specific subtopics within? 
It can be a good idea to make the mentions of subtopics in the intro text the same as the titles the audience finds below.

Author: It's a developer audience, so I think it's sufficient.

Author: Their goal is to flatten the subarrays and collections of key-value pairs...

within DataWeave objects, arrays, and subarrays.
//LINK TO DW 1.0 EXAMPLES:
include::partial$dataweave1-links.adoc[tag=dataweave1Examples]

> === Flatten Subarrays with Key-Value Pairs into an Array of Objects

Reviewer: Any reason to go from a Level-1 title to a Level-3, skipping the second level?

This example shows how `flatten` acts on key-value pairs of the input array
defined by the variable `arrayOne`. Notice that every key-value pair in the
array becomes a separate DataWeave object.

The example uses this function:

> * `flatten` to move the elements from the subarrays to the parent array,

Reviewer: List format for just one item?
Suggest converting this into a plain second paragraph, or merging it into the previous.

Author: This is to sync with format of the other pages, which often cover multiple functions.

eliminate the subarrays, and covert all key-value pairs into a list of objects
within the parent array.

.DataWeave:
[source,dataweave,linenums]
----
%dw 2.0
var arrayOne = [
                 [
                    "keyOne" : 1,
                    "keyTwo" : 2
                 ],
                 [
                    "keyThree" : 3,
                    "keyFour" : 4,
                    "keyFive" : 5
                 ],
                 "keySix" : 6
               ]
output application/json
---
flatten(arrayOne)
----

.Output JSON:
[source,json,linenums]
----
[
  {
    "keyOne": 1
  },
  {
    "keyTwo": 2
  },
  {
    "keyThree": 3
  },
  {
    "keyFour": 4
  },
  {
    "keyFive": 5
  },
  {
    "keySix": 6
  }
]
----

> === Flatten Combined Arrays

Reviewer: ditto re. Level-3 title?

> Both DataWeave examples return an array that combines and flattens the

Reviewer: ```suggestion
Like the previous DataWeave example, the following returns an array that combines and flattens the
```

Reviewer: As I get further into this topic set, I become confused about "both examples" in light of the fact that there are actually many examples. Do we mean "both" as in the two subtopics?
GLOBALLY search for "example" and evaluate whether the context makes clear what we mean when we refer to previous or next or first or second, and so on.

> elements from two arrays. The second example applies the selector `fruit`

Reviewer: ```suggestion
elements from two arrays. This second example applies the selector `fruit`
```

to select only the values of the key `fruit` and to exclude other values in
the array.

> The next example uses:

Reviewer: ```suggestion
This example uses:
```

* `++` to combine the objects of two arrays (`arrayOne` and `arrayTwo`) into
a single array.

> * `flatten` to flatten the combined array into a list of key-value pairs.

Reviewer: Remove periods from the two list items, because they are not themselves complete sentences.

And yes, I do think it is OK that this is a list (for two items) and the item in the earlier text NOT be in a list alone.

[[ex_1]]
.DataWeave:
[source,dataweave,linenums]
----
%dw 2.0
var arrayOne = [
                  [
                     "fruit" : "orange",
                     "fruit" : "apple"
                  ],
                  [
                     "fruit" : "grape",

>                      "notfruit" : "something else"

Reviewer: Instead of "something else", how about something like "pizza" or "hamburger" or "noodles"?

                  ]
               ]
var arrayTwo = [
                  [
                  	{  "fruit" : "kiwi" }
                  ],
                  "fruit" : "strawberry",
                  "fruit" : "plum",
                  { "fruit" : "banana" },
                  "notfruit" : "something else"
               ]
output application/json
---
flatten(arrayOne ++ arrayTwo)
----

.Output JSON:
[source,json,linenums]
----
[
  { "fruit": "orange" },
  { "fruit": "apple" },
  { "fruit": "grape" },
  { "notfruit": "something else" },
  { "fruit": "kiwi" },
  { "fruit": "strawberry" },
  { "fruit": "plum" },
  { "fruit": "banana" },
  { "notfruit": "something else" }
]
----

The only difference between the <<ex_1, previous example>> and the

> next is addition of the `.fruit` selector to the body expression

Reviewer: ```suggestion
following example is the addition of the `.fruit` selector to the body expression
```

                                      "fruit" : "watermelon",
                                      "notfruit" : "something else"
                                   }
                                },
                                {
                                   fruit : "cantaloupe",
                                   "notfruit" : "something else"
                                }
                             ]
                           },
                           {
                              fruit : "honeydew",
                              "notfruit" : "something else"
                           }
                        ]
                    },
                    { parentTwo:
                     	  [
                     	  	 fruit : "cherry",
                     		   "notfruit" : "something else"
                     	  ]
                    }
                 ]
output application/json
---
flatten(arrayOne ++ arrayTwo ++ arrayThree)..*fruit
----

.Output JSON:
[source,json,linenums]
----
[
  "orange",
  "apple",
  "grape",
  "kiwi",
  "strawberry",
  "plum",
  "banana",
  "watermelon",
  "cantaloupe",
  "honeydew",
  "cherry"
]
----

> === Flatten Subarrays into the Parent Array

Reviewer: ditto re. Level-3

This example shows how the `flatten` function acts on a variety of data types,
including numbers, DataWeave objects, subarrays, a string, a `null` value, and
and a key-value pair. Notice that elements of the subarrays become elements of
the parent array, and the subarrays are no longer present.

The example uses this function:

> * `flatten` to flatten the subarrays into their parent arrays.

Reviewer: ditto re. not using list format for just one item

.DataWeave:
[source,dataweave,linenums]
----
%dw 2.0
var myArray = [
                 1,
                 [2,3],
                 { a : "b"},
                 "my string",
                 [ [4,5], { c : "d"}, 6 ],
                 null,
                 "e" : "f"
              ]
output application/json
---
flatten(myArray)
----

.Output JSON:
[source,json,linenums]
----
[
  1,
  2,
  3,
  {
    "a": "b"
  },
  "my string",
  [
    4,
    5
  ],
  {
    "c": "d"
  },
  6,
  null,
  {
    "e": "f"
  }
]
----

== See Also

> * xref:dw-core-functions-flatten.adoc[flatten]

Reviewer: This looks incomplete.... Just *flatten ?

> * xref:dataweave-cookbook-define-function-to-flatten-list.adoc[Define a Function that Flattens Data in a List]

Reviewer: ```suggestion
* xref:dataweave-cookbook-define-function-to-flatten-list.adoc[Define a Function That Flattens Data in a List]
```
